test/error_messages.coffee

```diff
@@ -1122,15 +1122,15 @@ test "cannot export * without a module to export from", ->
   '''
 
 test "imports and exports must be top-level", ->
```

Comment:
This is the only one (at least from `test/error_messages.coffee`) that still isn't being thrown during AST generation (but in theory could) - the logic for throwing an error on "non-top-level" imports/exports is based on `o.indent`, which doesn't get set during AST generation

Seems like there could be a different mechanism not based on indent level to detect when you're not at the top level in general?

Reply:
Does `if o.level > LEVEL_TOP` not work?

Reply:
I think that wouldn't work since `level` really only reflects the "compilation level" *within a statement*

ie block-level statements always get compiled at `LEVEL_TOP`, eg here the import statement gets compiled at `LEVEL_TOP`:
```
if foo
  import { bar } from 'lib'
```

But it does seem like there could be a separate concept of "program top-level" (meaning I guess "a statement in the root `Block`"?)

It seems like that could get set on `o`. If we ended up tracking block scopes eg to be able to generate `const`/`let` where possible per your other comment, it seems like that would suffice

Reply:
> It seems like that could get set on `o`. If we ended up tracking block scopes eg to be able to generate `const`/`let` where possible per your other comment, it seems like that would suffice

Is tracking block scopes something that’s easy to add? If so, absolutely let’s do that, I’d love to start outputting `let`/`const` 😄 

Does a node know its parents? Obviously it knows its children, but if it also can easily find its parents then you can see if there are any `Block`s among its parents.

This is also not the most important check. The runtime will throw if anyone tries to put an `import` in not-top-level code in executable JavaScript (as opposed to JS further transpiled by some other build tool). We could just stop checking for this in CoffeeScript entirely, or check only on JS output with a comment saying why it’s hard to do this check on AST output, and leave it at that.

Reply:
> Is tracking block scopes something that’s easy to add? If so, absolutely let’s do that, I’d love to start outputting `let`/`const` 😄

I'm not sure, but I'd be up for doing a deeper dive on how scopes work sometime soon

> Does a node know its parents?

Ya I was thinking something similar. Currently I don't think there's any way to walk "up" the node class object tree. And I can't picture a simple way to add "parent links" to enable that

> This is also not the most important check

Agreed, I think it's fine for this to remain something that only gets flagged during compile-to-JS for the time being

> or check only on JS output with a comment saying why it’s hard to do this check on AST output

Sure, added a comment to `ModuleDeclaration::checkScope()`